taf/testlib/common/loops.py

```diff
@@ -0,0 +1,303 @@
+
+This program is distributed in the hope it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
+more details.
+
+@file  loops.py
+
+@summary  flexible and extensible conditional loops helpers.
+"""
+
+import time
+from contextlib import suppress
+from collections import OrderedDict
+
+
+class CondLoopException(Exception):
+    pass
+
+
+class CondLoopBase(object):
+    """
+    """
+    END_FAIL = False
+    END_SUCCESS = True
+    KEEP_LOOPING = None
+
+    def __init__(self, end_fail=False, end_success=True, keep_looping=None):
+        super(CondLoopBase, self).__init__()
+
+        self.END_FAIL = end_fail
+        self.END_SUCCESS = end_success
+        self.KEEP_LOOPING = keep_looping
+
+        self.end_conds_fail = []
+        self.end_conds_succ = []
+        self.gen_iter = None
+        self.loops_counter = 0
+
+    def end_conditions_met(self):
+        """
+        Override this method. Expected behavior upon returns from this method:
+            True/False - terminate with success/failure
+            None - keep looping
+        """
+        raise NotImplementedError
+
+    def loops_gen(self):
+        res = self.end_conditions_met()
+        while res is self.KEEP_LOOPING:
+            yield res
+            res = self.end_conditions_met()
+        yield res
+
+    def add_fail_condition(self, cond):
+        self.end_conds_fail.append(cond)
+
+    def add_success_condition(self, cond):
+        self.end_conds_succ.append(cond)
+
+
+class CondLoopImpl(CondLoopBase):
+    """
+    """
+    OPTIONS = OrderedDict()
+
+    @classmethod
+    def register_option(cls, option=None):
+        """
+        Register a keyword option for the class in `cls`
+        """
+        if not option:
+            option = cls.OPTION
+        if option in cls.OPTIONS:
+            raise CondLoopException("Conflict at '{}'".format(option))
+        cls.OPTIONS[option] = cls
+
+    def __init__(self):
+        """
+        @brief   Loop a routine over until it has passed or until an end condition has been met
+        """
+        super(CondLoopImpl, self).__init__()
+        self.gen_iter = self.loops_gen()
+        self.last_loop_res = None
+
+    def _before_next(self):
+        pass
+
+    def _after_next(self):
+        if self.last_loop_res is None:
+            self.loops_counter += 1
+
+    def __next__(self):
+        self._before_next()
+        self.last_loop_res = next(self.gen_iter)
+        self._after_next()
+        return self.last_loop_res
+
+    def __iter__(self):
+        return self
+
+    def __call__(self, *args, **kwargs):
+        return self.run(*args)
+
+    def run(self, *conds):
+        self.end_conds_succ.extend(conds)
+        return self._run()
+
+    def _run(self):
+        status = self.KEEP_LOOPING
+        while status is self.KEEP_LOOPING:
```

Comment:
Maybe you meant `dropwhile` instead of `takewhile`.

Im not sure what the second argument to the `next` should be, if any? Might as well just error out since that case of exhausting the iterator after the last element was still `self.keep_looping` "should never happen"?